docs/design-proposals/multiple-statefulsets.md

```diff
@@ -0,0 +1,201 @@
+# Multiple statefulsets support
+
+## Background
+
+Now, TiDB Operator create only one statefulset for one PD/TiKV/TiDB component separately. This works very well in most cases.
+
+But it can't meet or hardly meet these requests:
+
+* When using local volume, only horizontal scaling is guaranteed, vertical scaling may fail if there are not enough resources on the node. We need vertical scaling anyway
+* Users can't specify an exact number of pods in a given AZ. For example deploy 3 TiKV in AZ-1, and 2 TiKV in AZ-2
+* For on-premise deployment, k8s nodes may not have the same size. To make most of the resources, it's better to allow different profiles for a single cluster.
+* One TiKV peer use different resources(CPU/Memory/Storage) from other peers in the same TiDB cluster
+* No isolation between different tidb-servers, a big AP query may affect TP query. The worst case is that AP query cause a tidb-server OOM, and this tidb-server is also serving TP request
+* No support offline the specific peers
+* No support deploy TiDB Operator on multiple K8s clusters
+
+Using local persistent volume, once deployed the PD and TiKV pod has to be on the node all its lifecycle, migrating it to other nodes with larger resources would be hard. However if we use multiple statefulsets, we can do the migration on the fly. This is useful when there is a hot region on a TiKV that can hardly split.
+
+So the TiDB Operator should support creating multiple statefulsets for one component to achieve all these requests:
+
+* Add a new statefulset with bigger resources to scale TiDB cluster vertically
+* Different statefulsets can use different AZs or different resources and any other attributes
+* Can offline any peers when all the statefulsets’s replicas is 1
+* Deploy different statefulsets to different K8s clusters to support multiple K8s clusters(need further research)
+
+## Proposal
+
+### API change
+
+Users can create multiple statefulsets in `charts/tidb-cluster/values.yaml`, for example: change one tikv(Figure 1) to more tikvs(Figure 2):
+
+Figure 1:
+
+``` yaml
+tikv:
+  replicas: 3
+  image: pingcap/tikv:v2.1.0
+  …
+```
+
+Figure 2:
+
+``` yaml
+tikv:
+  replicas: 3
+  image: pingcap/tikv:v2.1.0
+  …
+  extra:
```

Comment:
This should be renamed since `extra` is a potentially common key name. `sets` would be better.

Reply:
There are a `default` set and zero or more extra sets, so what about `extraSets` or `otherSets`

Reply:
Yes, that is better. I wan't sure if we would support having just sets.

Reply:
Yeah, that is the problem we just discussed. Removing the default statefulset would impact all the other sets that inherit the property from the default set.

Reply:
Just sets +1

```yaml
pd:
  - name: name-1
    replicas: 3
    ...
  - name: name-2
    replicas: 2
    ...
```

Reply:
@tennix @onlymellb  @xiaojingchen @gregwebs PTAL